experimental/Matrix/matrix.jl

```diff
@@ -0,0 +1,112 @@
+# Initilize GAP function, i.e. GAP reads the file matrix.g
+#
+function __init__()
+    GAP.Globals.Reread(GAP.GapObj(joinpath(Oscar.oscardir, "experimental", "Matrix", "matrix.g")))
+end
+
+################################################################################
+# Computes the JuliaMatrixRep of a MatrixElem in GAP
+#
+#    _wrap_for_gap(m::MatrixElem)
+#
+# Input: m::MatrixElem
+# Output: Returns the JuliaMatrixRep of m in GAP
+#
+# Example
+# ```jldoctest
+# julia> m = matrix(ZZ, [0 1 ; -1 0]);
+# julia> _wrap_for_gap(m)
+# GAP: <matrix object of dimensions 2x2 over Integer Ring>
+# ```
+# """
+function _wrap_for_gap(m::MatrixElem)
+    return GAP.Globals.MakeJuliaMatrixRep(m)
+end
+
+
+################################################################################
+# Computes a GAP Group where the generators are the JuliaMatrixRep of matrices.
+# Moreover, a nice morphism from G into a GAP matrix group over a finite field G2 is constructed
+# such that calculations in G can be handled by G2 automatically.
+#
+#     MatrixGroup(matrices::Vector{<:MatrixElem{T}}) where T <: Union{fmpz, fmpq, nf_elem}
+#
+# Input: matrices::Vector{<:MatrixElem{T}} where T <: Union{fmpz, fmpq, nf_elem}
+# Output: GAP group generated by the JuliaMatrixReps of matrices
+#
+# Example
+# ```jldoctest
+# julia> m1 = matrix(QQ, [0 1 ; -1 0]);
+# julia> m2 = matrix(QQ, [ -1 0; 0 1]);
+# julia> MatrixGroup([m1,m2])
+# GAP: <group with 2 generators>
+# ```
+# """
+function MatrixGroup(matrices::Vector{<:MatrixElem{T}}) where T <: Union{fmpz, fmpq, nf_elem}
+       @assert !isempty(matrices)
+    
+       # One should probably check whether all matrices are n by n (and invertible
+       # and such ...)
+
+       K = base_ring(matrices[1])
+       if K isa FlintIntegerRing
+          K = QQ
+       end
+       for mat in matrices
+            if !(K == mat.base_ring)
+                error("Matrices are not from the same base ring")
+            end
+       end
+       n = nrows(matrices[1])
+
+       Fq, matrices_Fq, OtoFq = Oscar.good_reduction(matrices, 2)
+
+       G = Oscar.MatrixGroup(n, Fq, matrices_Fq)
+       N = order(G)
+       if !isdivisible_by(Hecke._minkowski_multiple(K, n), N)
+          error("Group is not finite")
+       end
+
+       #G_to_fin_pres = GAP.Globals.IsomorphismFpGroupByGenerators(G.X, GapObj([ g.X for g in gens(G) ]))
+       #F = GAP.Globals.Range(G_to_fin_pres)
+       #rels = GAP.Globals.RelatorsOfFpGroup(F)
+
+       #gens_and_invsF = [ g for g in GAP.Globals.FreeGeneratorsOfFpGroup(F) ]
+       #append!(gens_and_invsF, [ inv(g) for g in GAP.Globals.FreeGeneratorsOfFpGroup(F) ])
+       #matrices_and_invs = copy(matrices)
+       #append!(matrices_and_invs, [ inv(M) for M in matrices ])
+       #for i = 1:length(rels)
+       #   M = GAP.Globals.MappedWord(rels[i], GapObj(gens_and_invsF), GapObj(matrices_and_invs))
+       #   if !isone(M)
+       #      error("Group is not finite")
+       #   end
+       #end
```

Comment:
We actually do need this code, to verify that the map given by mapping from the group over the number field to the group over the finite field is an isomorphism

Reply:
Right! I thought about this at breakfast too but forgot it again. Thanks!